src/ns-support.c

```diff
@@ -0,0 +1,406 @@
+	if (flock(group->lock_fd, LOCK_UN) < 0) {
+		die("cannot release lock for namespace group %s", group->name);
+	}
+	debug("released lock for namespace group %s", group->name);
+}
+
+void sc_create_or_join_ns_group(struct sc_ns_group *group)
+{
+	// Open the mount namespace file.
+	char mnt_fname[PATH_MAX];
+	must_snprintf(mnt_fname, sizeof mnt_fname, "%s%s", group->name,
+		      SC_NS_MNT_FILE);
+	int mnt_fd __attribute__ ((cleanup(sc_cleanup_close))) = -1;
+	mnt_fd =
+	    openat(group->dir_fd, mnt_fname,
+		   O_CREAT | O_RDONLY | O_CLOEXEC | O_NOFOLLOW, 0600);
+	if (mnt_fd == -1) {
+		die("cannot open mount namespace file for namespace group %s",
+		    group->name);
+	}
+	// attempt to join an existing group
+	debug
+	    ("attempting to re-associate the mount namespace with the namespace group %s",
+	     group->name);
+	if (setns(mnt_fd, CLONE_NEWNS) == 0) {
+		debug
+		    ("successfully re-associated the mount namespace with the namespace group %s",
+		     group->name);
+		return;
+	}
+	// Anything but EINVAL is an unexpected error.
+	//
+	// EINVAL is simply a sign that the file we've opened is not a valid
+	// namespace file descriptor. One potential case where this can happen is
+	// when another snap-confine tried to initialize the namespace but was
+	// killed before it managed to complete the process.
+	if (errno != EINVAL) {
+		die("cannot re-associate the mount namespace with namespace group %s", group->name);
+	}
+	debug
+	    ("cannot re-associate the mount namespace with namespace group %s, falling back to initialization",
+	     group->name);
+	// Create a new namespace and ask the caller to populate it.
+	// For rationale of forking see this:
+	// https://lists.linuxfoundation.org/pipermail/containers/2013-August/033386.html
+	//
+	// The eventfd created here is used to synchronize the child and the parent
+	// processes. It effectively tells the child to perform the capture
+	// operation.
+	group->event_fd = eventfd(0, EFD_CLOEXEC);
+	if (group->event_fd == -1) {
+		die("cannot create eventfd for mount namespace capture");
+	}
+	debug("forking support process for mount namespace capture");
+	pid_t pid = fork();
+	debug("forked support process has pid %d", (int)pid);
+	if (pid == -1) {
+		die("cannot fork support process for mount namespace capture");
+	}
+	if (pid == 0) {
+		// This is the child process which will capture the mount namespace.
+		//
+		// It will do so by bind-mounting it the SC_NS_MNT_FILE after
+		// the parent process calls unshare() and finishes setting up the
+		// namespace completely.
+#ifdef HAVE_APPARMOR
+		// Change the hat to a sub-profile that has limited permissions
+		// necessary to accomplish the capture of the mount namespace.
+		debug
+		    ("changing apparmor hat of the support process for mount namespace capture");
+		if (aa_change_hat("mount-namespace-capture-helper", 0) < 0) {
+			die("cannot change apparmor hat of the support process for mount namespace capture");
+		}
```

Comment:
Why `aa_change_hat()` instead of `aa_change_profile()`? This seems like a one way operation for the child so `aa_change_profile()` may be better. Also, I think you can drop `-helper` since it implies a helper binary (at least to me anyway).